services/src/main/java/org/keycloak/services/managers/UserSessionManager.java

```diff
@@ -77,43 +78,42 @@ public UserSessionModel findOfflineUserSession(RealmModel realm, String userSess
     }
 
     public Set<ClientModel> findClientsWithOfflineToken(RealmModel realm, UserModel user) {
-        List<UserSessionModel> userSessions = kcSession.sessions().getOfflineUserSessions(realm, user);
-        Set<ClientModel> clients = new HashSet<>();
-        for (UserSessionModel userSession : userSessions) {
-            Set<String> clientIds = userSession.getAuthenticatedClientSessions().keySet();
-            for (String clientUUID : clientIds) {
-                ClientModel client = realm.getClientById(clientUUID);
-                clients.add(client);
-            }
-        }
-        return clients;
+        return kcSession.sessions().getOfflineUserSessionsStream(realm, user)
+                .flatMap(userSession -> userSession.getAuthenticatedClientSessions().keySet().stream())
+                .map(clientUUID -> realm.getClientById(clientUUID))
+                .collect(Collectors.toSet());
     }
 
+    @Deprecated
     public List<UserSessionModel> findOfflineSessions(RealmModel realm, UserModel user) {
-        return kcSession.sessions().getOfflineUserSessions(realm, user);
+        return this.findOfflineSessionsStream(realm, user).collect(Collectors.toList());
+    }
+
+    public Stream<UserSessionModel> findOfflineSessionsStream(RealmModel realm, UserModel user) {
+        return kcSession.sessions().getOfflineUserSessionsStream(realm, user);
     }
 
     public boolean revokeOfflineToken(UserModel user, ClientModel client) {
         RealmModel realm = client.getRealm();
 
-        List<UserSessionModel> userSessions = kcSession.sessions().getOfflineUserSessions(realm, user);
-        boolean anyRemoved = false;
-        for (UserSessionModel userSession : userSessions) {
-            AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(client.getId());
-            if (clientSession != null) {
-                if (logger.isTraceEnabled()) {
-                    logger.tracef("Removing existing offline token for user '%s' and client '%s' .",
-                            user.getUsername(), client.getClientId());
-                }
-
-                clientSession.detachFromUserSession();
-                persister.removeClientSession(userSession.getId(), client.getId(), true);
-                checkOfflineUserSessionHasClientSessions(realm, user, userSession);
-                anyRemoved = true;
-            }
-        }
-
-        return anyRemoved;
+        AtomicBoolean anyRemoved = new AtomicBoolean(false);
+        kcSession.sessions().getOfflineUserSessionsStream(realm, user).collect(Collectors.toList())
```

Comment:
I don't think it would because after you map it you lose the reference to the userSession that is used in the forEach block. That is, the forEach in your suggestion has only a reference to a clientSession, the userSession is gone by then.

Reply:
I workaround it by `UserSessionModel userSession = clientSession.getUserSession();` inside the `forEach` block. But anyway the pros of my suggestion are disputable :) 
Whatever works better for you.

Reply:
Ugh, missed that. I guess your approach works then, which at least makes the scope of the forEach block smaller. I'll give it a try.